### PR TITLE
Report an error when an export function does not map to a vm.func.

### PR DESCRIPTION
**Pull request by ScottTodd**
Fixes https://github.com/iree-org/iree/issues/4245

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/VM/Target/Bytecode/BytecodeModuleTarget.cpp [hunk 1]
```diff
@@ -460,6 +460,15 @@ static LogicalResult buildFlatBufferModule(
         return iree_vm_ImportFunctionDef_end(fbb);
       }));
 
+  // Check that all export ops point to valid functions.
+  for (auto &exportFuncOp : exportFuncOps) {
+    if (!symbolTable.lookup<IREE::VM::FuncOp>(exportFuncOp.function_ref())) {
+      return moduleOp.emitError()
+             << "vm.func op named '" << exportFuncOp.function_ref()
+             << "' not found for export";
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/VM/Target/Bytecode/BytecodeModuleTarget.cpp:468]: I tried to add a test for this, but `iree-compile` does not recognize the `--verify-diagnostics` flag needed for `// expected-error` tests (such as in https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Dialect/VM/Target/Bytecode/test/module_encoding_smoke.mlir).

**Comment by ScottTodd** [comment]
> this should be a verifier on vm.export instead - then it can verify all over the place instead of just in this target area
> 
> to do this the vm.export op should implement SymbolUserOpInterface and a verify method like this: [5560567#diff-ddbb48201ea18ae0e968542232cca385a0c68a1d328121289070ba4d927cbf45R882](https://github.com/iree-org/iree/commit/55605679738050ce9c0c8431af34c6e1dc59a308#diff-ddbb48201ea18ae0e968542232cca385a0c68a1d328121289070ba4d927cbf45R882) (well, actually returning an error instead of that that does, of course :)

Done! PTAL :)

**Review by benvanik — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/VM/IR/VMOps.cpp [hunk 2]
```diff
@@ -201,6 +201,16 @@ void ExportOp::build(OpBuilder &builder, OperationState &result,
   result.attributes.append(attrs.begin(), attrs.end());
 }
 
+LogicalResult ExportOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  Operation *op = getOperation();
+  if (!symbolTable.lookupNearestSymbolFrom<IREE::VM::FuncOp>(
+          op, function_refAttr())) {
+    return op->emitError() << "vm.func op named '" << function_ref()
+                           << "' not found for export";
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/VM/IR/VMOps.cpp:209]: Hmm... a few emitc test failures

https://source.cloud.google.com/results/invocations/108099f2-3a6a-49b6-bd28-4a1438424f65/targets;sort=name/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-swiftshader%2Fpresubmit/log

```
FAILED: samples/static_library/simple_mul_c_module.h samples/static_library/simple_mul_c_module.o samples/static_library/simple_mul_emitc.h /tmpfs/src/github/iree/build/samples/static_library/simple_mul_c_module.h /tmpfs/src/github/iree/build/samples/static_library/simple_mul_c_module.o /tmpfs/src/github/iree/build/samples/static_library/simple_mul_emitc.h
cd /tmpfs/src/github/iree/build/samples/static_library && /tmpfs/src/github/iree/build/tools/iree-compile --iree-mlir-to-vm-c-module --iree-hal-target-backends=dylib-llvm-aot --iree-llvm-link-embedded=false --iree-llvm-link-static --iree-llvm-static-library-output-path=simple_mul_c_module.o /tmpfs/src/github/iree/samples/static_library/simple_mul.mlir -o simple_mul_emitc.h
/tmpfs/src/github/iree/samples/static_library/simple_mul.mlir:1:1: error: vm.func op named 'module_simple_mul_export_shim' not found for export
func.func @simple_mul(%arg0: tensor<4xf32>, %arg1: tensor<4xf32>) -> tensor<4xf32>
^
/tmpfs/src/github/iree/samples/static_library/simple_mul.mlir:1:1: note: see current operation: "vm.export"() {export_name = "simple_mul", function_ref = @module_simple_mul_export_shim, iree.abi.stub, ordinal = 0 : i32} : () -> ()
/tmpfs/src/github/iree/samples/static_library/simple_mul.mlir:0:0: error: failed during transform passes
```

Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/VM/IR/VMOps.cpp:209]: cc @marbre + @simon-camp 

https://github.com/iree-org/iree/pull/8057 added these "_export_shim" ops. I might want some help debugging why this new verifier is breaking under emitc.

Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/VM/IR/VMOps.cpp:209]: Perfect, thank you! Tests are passing now :)

**Pull request merged**